### DISCUSSION TITLE
Is there a way to access device camera?

### DISCUSSION THREAD

**Discussion by Kofituo** [emilk/egui#3457] [Q&A]
I want to start a short personal project to expand my knowledge in Rust and gui program. I want to basically create a rtc program. In doing so, I want to add a video call feature. Before I start, I want to confirm if eframe/egui supports live camera feature

**Comment by tom-kaufman** [comment]
egui wouldn't support this directly. it is just a GUI framework. But, you can draw whatever you'd like onto the GUI. You will just need to find another crate that lets you get the webcam feed. It might be tricky to turn the video feed into images so that they can be displayed in your GUI.

**Reply by Kofituo** [reply]
I was thinking of using the default imaging library provided by egui (after getting the webcam frames). Since this is an immediate mode gui, setting the image about 30 times per second shouldn't be an issue. Am I right?

**Reply by tom-kaufman** [reply]
Yes it should work.

**Reply by Kofituo** [reply]
Well, I wasn't able to get it to work with the default imaging library. Seems the library requires already processed images ie png, jpeg, etc. Something like this just doesn't work

```rust
    fn update (...) {
            let camera = self.camera.as_mut().unwrap();
            let frame = camera.frame().unwrap();
            let buffer = Buffer::new(
                Resolution::new(1920, 1080),
                &frame.buffer(),
                FrameFormat::YUYV,
            );

            let image = buffer.decode_image::<LumaFormat>().unwrap();
            let raw = DynamicImage::ImageLuma8(image).into_luma8();

            ui.image(ImageSource::Bytes {
                 uri: Cow::from("bytes://frame.png /* what do I pass as the extension? */"),
                 bytes: Bytes::Shared(Arc::from(raw.as_raw().clone())),
             });
```
The above code produces an error in the ui

Even if I try saving the file and then using it in the ui doesn't also work (probably not the most efficient way to do things but I still gave it a try)

```rust
      let image = buffer.decode_image::<LumaFormat>().unwrap();
      image.save("testing.png").unwrap();
      ui.image(ImageSource::Bytes {
            uri: Cow::from("bytes://frame.png"),
            bytes:    Bytes::Shared(Arc::from(std::fs::read("testing.png").unwrap().clone())),
                });
```

The above shows only the first image (which is blank, since it takes time for the camera to set up)

**Reply by tom-kaufman** [reply]
This is what I meant by, "It might be tricky to turn the video feed into images so that they can be displayed in your GUI."

I recently had to display some images from a network drive, that can't be included at compile time, in my egui app. I can't share the source with you, but let me give the tree of egui docs links I had to follow to get it working on my end.

We want to create an `Image` widget with [`Image::from_texture()`](https://docs.rs/egui/latest/egui/widgets/struct.Image.html#method.from_texture). So, we need a `texture: impl IntoSizedTexture>`. We'll use [`SizedTexture::from_handle(handle: &TextureHandle)`](https://docs.rs/egui/latest/egui/load/struct.SizedTexture.html#method.from_handle). The [`TextureHandle`](https://docs.rs/egui/latest/egui/struct.TextureHandle.html) docs page points use to use [`egui::Context::load_texture`](https://docs.rs/egui/latest/egui/struct.Context.html#method.load_texture).

Note this passage from the `load_texture()` docs:  

> Make sure to only call this ONCE for each image, i.e. NOT in your main GUI code. The call is NOT immediate safe. 

What I take this to mean, is that you should not be trying to re-load an image file into a texture once per frame. For your application, I think this is just indicating that this may be a performance bottleneck. One thing you might do is put a timer on whatever method is updating the texture that you want to display, so that you don't load the webcam image into a texture more than once every, say, 1/24th of a second. You might achieve this by putting the webcam refresh loop in a background thread, and sending the updated `TextureHandle` to your egui application state struct via an mpsc channel.

Anyway, now for loading the `TextureHandle`. The `load_texture()` docs points us to the `ColorImage::from_rgba_unmultiplied()` docs, which gives us an [example](https://docs.rs/egui/latest/egui/struct.ColorImage.html#example-using-the-image-crate) function for loading a `ColorImage` from bytes in memory or a file on disk.

Hopefully, with whatever webcam library you are using, you can transform the frame from the video into a byte slice, that could be written to a jpg or bmp or whatever, and be a valid image file. Then, try passing that byte slice to `load_image_from_memory()` from the example snippet.

Let me know how it goes!

**Reply by Kofituo** [reply]
While this works, the main issue is `ColorImage::from_rgba_unmultiplied()` basically requires the image to have been processed into a format such as png, jpeg, etc first before it's able to be displayed. There isn't a smooth way to display it as a continuous video

**Reply by Kofituo** [reply]
If there was a video library, that may have made the work a bit easier, I guess

**Reply by tom-kaufman** [reply]
At the end of the day, `egui` is just drawing triangles and textures onto the screen with OpenGL or whatever graphics backend is being used. You can't get around needing to turn your video feed into an image. You may have better luck constructing a `ColorImage` directly from your video feed, rather than doing video feed -> image file -> `ColorImage`. A `ColorImage` is just a vec of `Color32`. I don't know what you are using to collect your video feed, or what kind of information that gives you, but it should be doable to bypass constructing a JPG.

**Reply by Kofituo** [reply]
I'm using [nokhwa](https://crates.io/crates/nokhwa) library

**Reply by tom-kaufman** [reply]
It looks like you can pretty easily get a vector of pixels from that library (e.g. https://docs.rs/image/0.24.6/image/struct.ImageBuffer.html#method.pixels), which you could then almost directly use to construct a `ColorImage` for use in `egui`.

**Reply by Kofituo** [reply]
Here's what I've done so far
https://gist.github.com/Kofituo/292660231e614dc8d4c3708e5207031a

The image still don't get displayed and it seems it's extremely laggy. From the docs `load_texture` is not immediate safe. Yet, I can't seem to find a working alternative.

**Reply by tom-kaufman** [reply]
I took a crack at this, but this is as far as I got. It seems like using `ColorImage::as_raw` doesn't give use the same byte layout that `Bytes` is expecting.

```
use eframe::egui::load::{Bytes, SizedTexture};
use eframe::egui::{ColorImage, Context, ImageButton, ImageSource, Image};
use eframe::{egui, Frame, CreationContext};
use image::codecs::png::PngEncoder;
use image::{
    ColorType, DynamicImage, ImageBuffer, ImageEncoder, ImageFormat, Luma, Pixel,
    PixelWithColorType,
};
use nokhwa::pixel_format::{LumaAFormat, LumaFormat, RgbAFormat, RgbFormat};
use nokhwa::utils::{CameraIndex, FrameFormat, RequestedFormat, RequestedFormatType, Resolution};
use nokhwa::{Buffer, Camera};
use std::borrow::Cow;
use std::collections::{HashMap, HashSet};
use std::sync::mpsc::{Receiver, TryRecvError};
use std::sync::{mpsc, Arc};
use std::time::Duration;

fn main() {
    env_logger::init(); // Log to stderr (if you run with `RUST_LOG=debug`).

    let options = eframe::NativeOptions {
        initial_window_size: Some(egui::vec2(600.0, 800.0)),
        ..Default::default()
    };
    eframe::run_native(
        "Image Viewer",
        options,
        Box::new(|cc| {
            // This gives us image support:
            egui_extras::install_image_loaders(&cc.egui_ctx);
            Box::new(App::new(cc))
        }),
    )
    .unwrap();
    println!("progress")
}

struct App<'a> {
    image: Receiver<ImageBuffer<Luma<u8>, Vec<u8>>>,
    // image_bytes: Arc<[u8]>,
    image_source: Option<ImageSource<'a>>,
}

impl<'a> App<'a> {
    fn new(_cc: &CreationContext) -> App<'a> {

        let (sender, receiver) = mpsc::channel();
        std::thread::spawn(move || {
            println!("Background thread starting");
            let mut camera = App::set_up_camera();
            loop {
                let frame = camera.frame().unwrap();
                println!("Background task got a frame");

                let buffer = Buffer::new(
                    Resolution::new(1920, 1080),
                    &frame.buffer(),
                    FrameFormat::YUYV,
                );
                let image = buffer.decode_image::<LumaFormat>().unwrap();
                println!("Background task sending an image");
                sender.send(image).unwrap()
            }
        });
        // let image_source = ImageSource::Bytes {
        //     uri: "bytes://feed.png".into(),
        //     bytes: Bytes::Shared(image_bytes.clone()),
        // };
        println!("1E");
        Self {
            image: receiver,
            image_source: None,
        }
    }

    fn set_up_camera() -> Camera {
        let index = CameraIndex::Index(0);
        // request the absolute highest resolution CameraFormat that can be decoded to RGB.
        let requested =
            RequestedFormat::new::<RgbFormat>(RequestedFormatType::AbsoluteHighestResolution);
        // make the camera
        let mut camera = Camera::new(index, requested).unwrap();
        camera.open_stream().unwrap();
        camera
    }
}

impl<'a> eframe::App for App<'a> {
    fn update(&mut self, ctx: &Context, frame: &mut Frame) {
        egui_extras::install_image_loaders(ctx);
        
        if self.image_source.is_none() {
            let img = ColorImage::example();
            println!("1A");
            let image_bytes: Arc<[u8]> = unsafe {
                let slice = img.as_raw().as_ref();
                println!("1B");
                Arc::from_raw(slice)
            };
            println!("1C");
            let image_bytes = Bytes::Shared(image_bytes);
            println!("1D");
            ctx.include_bytes("bytes://feed.png", image_bytes);
            let image_source = ImageSource::Uri("bytes://feed.png".into());
            self.image_source = Some(image_source);
        }
        egui::CentralPanel::default().show(ctx, |ui| {
            egui::ScrollArea::new([true, true]).show(ui, |ui| {
                let image = self.image.try_recv();
                println!("ok {}", image.is_ok());


                // Uncomment this and make it match whatever lines 92-106 end up looking like
                // match image {
                //     Ok(image) => {
                //         let im = ColorImage::from_gray(
                //             [image.dimensions().0 as _, image.dimensions().1 as _],
                //             &image.as_raw(),
                //         );
                //         println!("2A");
                //         let image_bytes: Arc<[u8]> = unsafe {
                //             let slice = im.as_raw().as_ref();
                //             println!("2B");
                //             Arc::from_raw(slice)
                //         };
                //         println!("2C");
                //         let image_source = ImageSource::Bytes {
                //             uri: "bytes://feed".into(),
                //             bytes: Bytes::Shared(image_bytes.clone()),
                //         };
                //         println!("2D");
                //         self.image_source = image_source;
                //     }
                //     Err(e) => {
                //         println!("error {:?}", e)
                //     }
                // }


                // let texture: &egui::TextureHandle = self.texture.get_or_insert_with(|| {
                //     // Load the texture only once.
                //     // ui.ctx()
                //     //     .load_texture("my-image", self.image_bytes.clone(), Default::default())
                // });

                ui.label("Hello");
                if let Some(source) = &self.image_source {
                    println!("3A");
                    let image = Image::new(source.clone());
                    println!("3B");
                    ui.add(image);
                    println!("3B");
                }
            });
        });
    }
}
```

Edit: And here's what my Cargo.toml looks like:
```
[package]
name = "webcamthing"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
eframe = "0.23.0"
egui_extras = { version = "0.23.0", features = ["all_loaders"] }
env_logger = "0.10.0"
image = { version = "0.24.7", features = ["jpeg", "png"] }
nokhwa = { version = "0.10.4", features = ["input-native"] }
```
